labs/lab03.md

```diff
@@ -0,0 +1,52 @@
+# Lab 3: Testing
+
+In this lab you will get familiar with different testing strategies and practice with writing good unit tests in Java and TypeScript.
+
+## Concepts
+**Structural testing vs Specification testing:** Two distinct types of unit testing are specification-based (historically: black-box) testing and structural (white-box) testing. Specification testing is when you write high-level test cases against the specification of a program, without using (or testing) the specific underlying implementation of the program. In structural testing, your tests are instead tailored to the specific underlying implementation of the code you are testing.
+
+**Code coverage:** Coverage is the amount of code that has been executed by the tests. There are several coverage metrics, but one common metric line coverage is simply the percentage of lines (or statements) that have been executed by your unit tests. 
+
+## Deliverables
+
+- [ ] Write tests for LinkedIntQueue in TypeScript and ArrayIntQueue in Java and execute them from within the IDE.
+- [ ] Achieve 100% line coverage for ArrayIntQueue and show coverage results in the IDE or as HTML report. 
+- [ ] Compare structural testing vs specification testing and analyze tradeoffs. 
+
+
+## Instructions
+
+### Setup
+Fork and clone the repo from https://github.com/CMU-17-214/f23-rec03.
+The code is set up with JUnit and ts-jest. You can use your IDE to open the folder of each language and run the existing test with the command in the readme.txt. You can also see Appendix 1 for how to run tests in IDE. 
+
+
+### Testing
+Write additional tests for code we've provided and use a code coverage tool to measure the coverage of your unit tests. 
+
+For LinkedIntQueue you will need to use specification-based testing, because the source code for LinkedIntQueue is not provided. Your tests will be based on the specified functionality of the IntQueue interface rather than the code of the underlying LinkedIntQueue implementation. 
```

Comment:
Add subsections/headings **TypeScript: ** and **Java:** to make clear what is to be done in which language?